src/condor_utils/compat_classad.h

```diff
@@ -23,13 +23,88 @@
 #include "classad/classad_distribution.h"
 #include "classad_oldnew.h"
 
+
 using classad::ClassAd;
 
 class Stream;
 
 class MapFile; // forward ref
-
 class ClassAdFileParseHelper;
+class auto_free_ptr;
+
+class CompatFileLexerSource : public classad::FileLexerSource
+{
+public:
+	CompatFileLexerSource() = default;
+	CompatFileLexerSource(FILE* fp, bool close_file=false)
+		: classad::FileLexerSource(fp)
+		, _close_on_delete(close_file)
+	{}
+	CompatFileLexerSource(const CompatFileLexerSource &) = delete;
+	CompatFileLexerSource &operator=(const CompatFileLexerSource &) = delete;
+
+	virtual ~CompatFileLexerSource() {
+		if (_close_on_delete && _file) { fclose(_file); }
+		_file = nullptr;
+	}
+
+	void SetSource(FILE *file, bool close_file=false) {
+		// we don't expect to ever be called with a non-null _file here, but just in case
+		if (_close_on_delete && _file) { fclose(_file); }
+		_close_on_delete = close_file;
+		this->SetNewSource(file);
+	}
+
+	bool readLine(std::string & buffer, bool append=false);
+protected:
+	bool _close_on_delete{false};
+};
+
+class CompatStringViewLexerSource : public classad::StringViewLexerSource
+{
+public:
+	CompatStringViewLexerSource() = default;
+	CompatStringViewLexerSource(std::string_view sv, int offset=0) : classad::StringViewLexerSource(sv,offset) {}
+	virtual ~CompatStringViewLexerSource() = default;
+
+	bool readLine(std::string & buffer, bool append=false);
+	const char * data(size_t * cb=nullptr, int * off=nullptr) {
+		if (cb) *cb = _strview.size();
+		if (off) *off = _offset;
+		return _strview.data();
+	}
+};
+
+// this LexerSource makes a copy of the input string
+class CompatStringCopyLexerSource : public CompatStringViewLexerSource
+{
+public:
+	// constructors that steal an auto_free_ptr
+	CompatStringCopyLexerSource(auto_free_ptr & that); 
+	CompatStringCopyLexerSource(auto_free_ptr & that, size_t len);
+	// constrctor that copies a string pointer
+	CompatStringCopyLexerSource(const char * ptr, size_t len, size_t extra=0) 
+		: _strcopy(strdup_len(ptr, len, extra)) {
+		SetNewSource(std::string_view(_strcopy, len+extra), 0);
+	}
```

Comment:
The extra argument is a major foot-gun (uninitialized memory in the buffer) and not used anywhere. We should get rid of it.

Reply:
we could zero initialize.  but its required if you want to append something without first making a copy of the stringview that is being copied.   yes it's not being used yet, but most of the uses of this are in the future.

Reply:
Zero-initializing would address my main concern. But the whole premise seems wrong to me. Hey, LexerSource, here's a buffer that you own/control that I'm going to modify later, potentially while you're reading through it. Do you have a concrete example where the caller knows the extra space it will need but can't fill in the contents itself before giving the buffer to the LexerSource?

Reply:
Added zero init of extra data and zero term of initial data (which is a bit wrong, but can be worked around in the unit test). See the unit test for a plausible example.